File: src/rest-scratch-node-express/external/data-scratch-node-library/ts-data-scratch/machine_learning.ts
// data splitting

function  split_data(data:Array<Array<number>>, prob:number) {
    // split data into fractions [prob, 1 - prob]
    var results = { "train":[], "test":[] }
    for (var i = 0; i < data.length; i += 1 ) {
        var row=data[i]
        if (Math.random() < prob) {
            results["test"].push(row)
        } else {
            results["train"].push(row)
        }
    }
    return results
}

function train_test_split(x:Array<Array<number>>, y:Array<number>, test_pct:number) {
    var results = {
        "x_train":[],
        "y_train":[],
        "x_test":[],
        "y_test":[]
    }
    for (var i = 0; i < x.length; i += 1 ) {
        var x_i = x[i]
        var y_i = y[i]
        if (Math.random() < test_pct) {
            results["x_test"].push(x_i)
            results["y_test"].push(y_i)
        } else {
            results["x_train"].push(x_i)
            results["y_train"].push(y_i)
        }        
    }
    return results
}

// correctness

function  accuracy(tp:number, fp:number, fn:number, tn:number) {
    var correct = tp + tn
    var total = tp + fp + fn + tn
    return correct / total
}

function  precision(tp:number, fp:number, fn:number, tn:number) {
    return tp / (tp + fp)
}

function  recall(tp:number, fp:number, fn:number, tn:number) {
    return tp / (tp + fn)
}

function  f1_score(tp:number, fp:number, fn:number, tn:number) {
    var p = precision(tp, fp, fn, tn)
    var r = recall(tp, fp, fn, tn)
    return 2 * p * r / (p + r)
}

File: src/data-scratch-cpp-library/dscpp/c11_machine_learning/machine_learning.cpp
// Machine Learning Implementation - C++ Data Science Library
// Port from TypeScript implementation

#include "machine_learning.h"
#include <random>
#include <stdexcept>

// Random number generator
static std::random_device rd;
static std::mt19937 gen(rd());

// Data splitting functions
SplitData split_data(const std::vector<std::vector<double>>& data, double prob) {
    // split data into fractions [prob, 1 - prob]
    SplitData results;
    std::uniform_real_distribution dis(0.0, 1.0);
    
    for (const auto& row : data) {
        if (dis(gen) < prob) {
            results.test.push_back(row);
        } else {
            results.train.push_back(row);
        }
    }
    return results;
}

TrainTestSplit train_test_split(const std::vector<std::vector<double>>& x, 
                               const std::vector<double>& y, 
                               double test_pct) {
    if (x.size() != y.size()) {
        throw std::invalid_argument("x and y must have the same length");
    }
    
    TrainTestSplit results;
    std::uniform_real_distribution dis(0.0, 1.0);
    
    for (size_t i = 0; i < x.size(); ++i) {
        if (dis(gen) < test_pct) {
            results.x_test.push_back(x[i]);
            results.y_test.push_back(y[i]);
        } else {
            results.x_train.push_back(x[i]);
            results.y_train.push_back(y[i]);
        }
    }
    return results;
}

// Performance metrics
double accuracy(double tp, double fp, double fn, double tn) {
    double correct = tp + tn;
    double total = tp + fp + fn + tn;
    return (total == 0) ? 0.0 : correct / total;
}

double precision(double tp, double fp, [[maybe_unused]] double fn, [[maybe_unused]] double tn) {
    double denominator = tp + fp;
    return (denominator == 0) ? 0.0 : tp / denominator;
}

double recall(double tp, [[maybe_unused]] double fp, double fn, [[maybe_unused]] double tn) {
    double denominator = tp + fn;
    return (denominator == 0) ? 0.0 : tp / denominator;
}

double f1_score(double tp, double fp, double fn, double tn) {
    double p = precision(tp, fp, fn, tn);
    double r = recall(tp, fp, fn, tn);
    double denominator = p + r;
    return (denominator == 0) ? 0.0 : 2.0 * p * r / denominator;
}
